Refactor inbound authentication with custom provider and handlers (#15056)
ballerina-platform/ballerina-lang: stdlib/auth/src/main/ballerina/auth/Module.md

## Module Overview

> This module provides a set of default authentication providers that can be extended to create new authentication providers. 

Reviewer: ```suggestion
This module provides a set of default authentication provider configurations that can be extended to create new authentication providers. 
```

> This module provides a set of default authentication providers that can be extended to create new authentication providers. 

Reviewer: ```suggestion
This module provides a set of default authentication provider configurations that can be extended to create new authentication providers. 
```

### Authentication Provider

> An authentication provider defines an authentication scheme that could be used to protect endpoints. The `auth:AuthStoreProvider` type acts as the interface for all the authentication providers. Any type of implementation, such as LDAP, JDBC, and file based, should be object-wise similar.

Reviewer: ```suggestion
An authentication provider defines an authentication scheme that could be used to authenticate endpoints. The `auth:AuthStoreProvider` acts as the interface for all the authentication providers. Any type of implementation such as LDAP, JDBC, and file-based should be object-equivalent.
```

> By default, there are three implementations of the `auth:AuthProvider`. They are; the `auth:ConfigAuthStoreProvider`,

Reviewer: ```suggestion
```

>  which authenticates based on usernames and passwords stored in a configuration file; the `auth:JWTAuthProvider`, which

Reviewer: ```suggestion
```

>  authenticates by validating a JWT; and finally the `auth:LdapAuthStoreProvider`, which authenticates based on the user

Reviewer: ```suggestion
```

>  credentials stored in an active directory or an LDAP.

Reviewer: ```suggestion
By default, there are three implementations of the `auth:AuthProvider`. They are:
1. The `auth:ConfigAuthStoreProvider`, which authenticates based on usernames and passwords stored in a configuration file.
2. The `auth:JWTAuthProvider`, which authenticates by validating a JWT.
3. The `auth:LdapAuthStoreProvider`, which authenticates based on the user credentials stored in an active directory or an LDAP.
```

> When creating a new authentication provider, there is a function that need to be implemented. 

Reviewer: ```suggestion
When creating a new authentication provider, there is a function that needs to be implemented. 
```

- `authenticate` : Authenticates the user based on a credential, which can be username/password, or a token such as JWT.

### Config Auth Store Provider

> `auth:ConfigAuthStoreProvider` is an implementation of the `auth:AuthProvider` interface, which uses the Ballerina configuration file

Reviewer: ```suggestion
The `auth:ConfigAuthStoreProvider` is an implementation of the `auth:AuthProvider` interface, which uses the Ballerina configuration file
```

>  to store usernames, passwords, scopes and the relevant associations.

Reviewer: ```suggestion
 to store usernames, passwords, scopes, and relevant associations.
```

A user is denoted by a section in the configuration file. The password and the scopes assigned to the user are denoted

>  as keys under the relevant user section as seen below. 

Reviewer: ```suggestion
 as keys under the relevant user section as shown below. 
```

### LDAP Auth Store Provider

> `auth:LdapAuthStoreProvider` is another implementation of the `auth:AuthProvider` interface, which connects to an active directory or an LDAP to retrieve the necessary user information to perform authentication and authorization.

Reviewer: ```suggestion
The `auth:LdapAuthStoreProvider` is another implementation of the `auth:AuthProvider` interface. This connects to an active directory or an LDAP, retrieves the necessary user information, and performs authentication and authorization.
```

### JWT Auth Provider

> `auth:JWTAuthProvider` is another implementation of the `auth:AuthProvider` interface, which authenticate by validating a JWT.

Reviewer: ```suggestion
The `auth:JWTAuthProvider` is another implementation of the `auth:AuthProvider` interface, which authenticates by validating a JWT.
```
